### ISSUE TITLE
program not working.

### ISSUE THREAD

**Issue by WiggyWire** [stardot/MatrixBrandy#17]
`10 REM - Market Day
20 MODE2
30 FOR X=0 TO 7
40 VDU19,X+8,X,0,0,0
50 NEXT
60 ROW=1100
70 REPEAT
80 SIZE=(60-ROW/20)*4
90 ROW=ROW-SIZE-2
100 FOR COLUMN=1 TO 15
110 MOVE COLUMN*SIZE,ROW-100
120 GCOLRND(255),0
130 PLOT0,SIZE,0
140 PLOT81,-SIZE,SIZE
150 PLOT1,SIZE,-SIZE
160 PLOT81,0,SIZE
170 NEXT
180 UNTIL ROW<=0`

Starting to see a problem as a bunch of listings I have tried don't seem to work at all as described in the book. I get not implementing something difficult like sound, but there seems to be things that were commonly used in the listings that just break a great deal of stuff that you will find in the magazines and books for the beeb.

**Comment by WiggyWire** [comment]
The book where basically nothing seems to be working is BBC Micro Programs in BASIC by Derrick Daines.

**Comment by soruk42** [comment]
Please check line 120 - looks like a typo. I get something out when I do GCOL0,RND(255)

**Comment by soruk42** [comment]
That's a good question. Many of the animation tricks, for instance, use palette changing, as on the BBC Micro this changes the actual colour of all pixels at once. In Matrix Brandy (and I believe, BB4W) this is not the case. Irrespective of the chosen screen mode, the display plane is a true-colour display, and pixels are plotted from the palette at the time they're plotted.

I could do something that searches and replaces pixels on a palette change, but that is also liable to problems as if two palette entries are set to the same value, the screen can't then tell them apart for future palette changes.

**Comment by WiggyWire** [comment]
Would it be easier to stick to Mode 7? that seems to be fairly complete. I just need to find some decent sources on how to program it. It seems to have been the red-headed step child of the modes, not much is written in the old books about it.

**Comment by jgharston** [comment]
Some very early BBC programs used out-of-range values to things like GCOL with the values being truncated, eg GCOL RND(blah) would be truncated to GCOL 0-7 with the defined effects for GCOL 0-4 and "odd" effects for GCOL 5-7 due to table overruns in the MOS.
Similar things happens with some music programs, where other than the channel the SOUND parameters are 8-bit numbers, and programmers blissfully did things like SOUND ch,thing*4... and it didn't matter that thing*4 was eg &0268 as the SOUND system just took the bottom byte. Later MOSes such as RISC OS though took the full 16-bit value, giving unexpected results.

**Comment by soruk42** [comment]
Earlier I wrote:
_I could do something that searches and replaces pixels on a palette change, but that is also liable to problems as if two palette entries are set to the same value, the screen can't then tell them apart for future palette changes._

I've got this working, by using the top byte of the 32 bits per pixel, in paletted modes, to hold the logical colour used. This is ignored by SDL, but has allowed me to track the pixel value even if two colours are the same. Thus, I now have VDU19 updating the screen with a palette change immediately on the latest commit.
